### ISSUE TITLE
db_info.txt read multiple times during normal run of Metalign

### ISSUE THREAD

**Issue by peterk87** [nlapier2/Metalign#21]
The `db_info.txt` file is read multiple times when running `metalign.py`:

https://github.com/nlapier2/Metalign/blob/8a7eba84ef7f62a40a28c3117967012a68a97bd0/select_db.py#L174

https://github.com/nlapier2/Metalign/blob/8a7eba84ef7f62a40a28c3117967012a68a97bd0/select_db.py#L41-L46

https://github.com/nlapier2/Metalign/blob/8a7eba84ef7f62a40a28c3117967012a68a97bd0/map_and_profile.py#L537

https://github.com/nlapier2/Metalign/blob/8a7eba84ef7f62a40a28c3117967012a68a97bd0/map_and_profile.py#L80-L85

It seems like a significant amount of time is spent reading the tab-delimited table into dicts since the file is very large (7.7GB with the default data). Maybe the file could be read only once and the data structure passed around as needed?

Maybe serializing the dicts to pickle or JSON format and reading the serialized data into Metalign would improve runtime?

**Comment by nlapier2** [comment]
Yes, that's an improvement that should be made at some point. I generally haven't focused on this much yet because, on my server, it only takes a minute or two to read in the dbinfo file, which is negligible compared to the overall runtime on large reads datasets, albeit irritating for small test datasets. But I will keep this on my to-do list. 

Thanks for the reminder on pickling, I will think about that. Have you seen before that this helps with read time for large data structures?

**Comment by peterk87** [comment]
> Have you seen before that this helps with read time for large data structures?

Yes, I've found that serialization and deserialization to and from a binary format is usually much faster especially for numerical or fixed size data. I usually try to serialize big array or tabular data in HDF5 format if possible since it's fairly ubiquitious and performant.

I forgot to mention that the memory usage of the dicts `acc2info` and `taxid2info` is *very* high. I'm seeing around 26 GB RSS memory usage after parsing `db_info.txt` with a slightly modified `get_acc2info` function (accepts a file path string instead of an `argparse.Namespace`) in a Jupyter Notebook. Reducing this down to a more reasonable level could allow Metalign to run on a regular laptop (or cost a lot less money to run in the cloud). 

With Pandas, I was able to get the memory usage down to 5GB and parsing the table in ~60s vs ~160s with `get_acc2info`. 

```python
import pandas as pd

col_dtypes = [('acc', str), 
              ('length', int), 
              ('taxid', 'category'), 
              ('lineage', 'category'), 
              ('taxid_lineage', 'category')]

df = pd.read_table(db_info, 
                   header=None, 
                   names=[x for x,y in col_dtypes], 
                   dtype={x:y for x,y in col_dtypes})

print(df.memory_usage(deep=True) / (1024**2))
# Index               0.000122
# acc              3165.778735
# length            352.320946
# taxid             182.685093
# lineage           179.263195
# taxid_lineage     182.682842
# dtype: float64
# ^ approx mem usage in MB of each column
```

Writing the Pandas DataFrame to an HDF5 took ~40s and created a 1.9GB file. Reading the same HDF5 file back into a DataFrame took ~35s.

```python
from pandas.testing import assert_frame_equal

df.to_hdf('dbinfo.h5', 'dbinfo', format='table')
df_from_hdf = pd.read_hdf('dbinfo.h5', 'dbinfo')

# both DataFrames should be equal
assert_frame_equal(df, df_from_hdf)
```

[Parquet](https://arrow.apache.org/docs/python/parquet.html) and [PyArrow](https://arrow.apache.org/docs/python) allow you to save the same DataFrame in a 229 MB file in 6s! Reading the Parquet file only took 22s. 

```python
df.to_parquet('dbinfo.parquet')
df_parquet = pd.read_parquet('dbinfo.parquet')
assert_frame_equal(df, df_parquet)
```

- [Parquet in Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-parquet)


***

What do you think of only parsing the lines from the `db_info.txt` that appear in the CMash results output file? That is, get a set of taxids from the CMash results and check to see if each line in the `db_info.txt` file contains a taxid in that set? 

With a fairly minor change to the source code:
```python
def acc2info_taxids(dbinfo: str, taxids: set) -> (dict, dict):
    acc2info = {}
    taxid2info = {}
    with(open(dbinfo, 'r')) as fh:
        fh.readline()
        for line in fh:
            acc, acclen, taxid, namelin, taxlin = line.strip().split('\t')
            # only continue parsing lines with a taxid of interest
            if not taxid in taxids:
                continue
            rank = get_taxid_rank(taxlin)
            if rank == 'strain' and acc != 'Unmapped':
                taxid += '.1'  # CAMI formatting specification
                taxlin += '.1'
            acclen = int(acclen)
            acc2info[acc] = [acclen, taxid, namelin, taxlin]
            if taxid in taxid2info:
                taxid2info[taxid][0] += acclen
            else:
                taxid2info[taxid] = [acclen, rank, namelin, taxlin]
    return acc2info, taxid2info
```

For a set of 1000 random taxids, it only takes 30s and a lot less memory (~100MB). 65s and ~5GB for the 1000 largest taxids (most accessions). 

Also, is it useful to have so much redundancy within certain taxa? The top 1000 taxa by number of entries consist of 19,294,247 out of 46,179,361 (41.8%) of the entries in the `db_info.txt`. Top 100 taxa are almost 20%. Maybe a CMash within taxa done to reduce redundancy?
